File: nomad/src/org/nomad/main/PatchLoader.java
/* Copyright (C) 2006 Christian Schneider
 * 
 * This file is part of Nomad.
 * 
 * Nomad is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * Nomad is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Nomad; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

/*
 * Created on Jan 23, 2006
 */
package org.nomad.main;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;

import javax.swing.SwingUtilities;

import org.nomad.dialog.NomadTaskDialog;
import org.nomad.dialog.TaskModel;
import org.nomad.patch.format.PatchConstructionException;
import org.nomad.patch.format.PatchBuilder;
import org.nomad.patch.format.PatchFile303;
import org.nomad.patch.ui.PatchUI;

class PatchLoader {

	private final Nomad nomad;
	private ArrayList<PatchFile> fileList = new ArrayList<PatchFile>();
	
	public PatchLoader(Nomad nomad) {
		this.nomad = nomad; }
	public void loadPatch(String file) {
		loadPatch(new String[]{file});
	}

	public void loadPatch(File[] files) {
		String[] sfiles = new String[files.length];
		for (int i=files.length-1;i>=0;i--)
			sfiles[i]=files[i].getAbsolutePath();
		
		loadPatch(sfiles);
	}

	private class PatchFile {
		String fileName;
		String patchName;
		public PatchFile(String fileName) {
			this.fileName = fileName;
			patchName = fileName.substring(0,fileName.lastIndexOf(".pch"));
			patchName = patchName.substring(patchName.lastIndexOf('/')+1);
		}
	}
	
	public void loadPatch(final String[] files) {
		for (String fileName : files) {
			fileList.add(new PatchFile(fileName));
		}

		NomadTaskDialog dlg = new NomadTaskDialog(new TaskModel(){
			
			int taskCount = fileList.size();
			int selection = nomad.getDocumentManager().getDocumentCount();
			
			public String getDescription() {
				return "Loading Nord Modular patch...";
			}

			public int getTaskCount() {
				return taskCount;
			}

			public String getTaskName(int taskIndex) {
				return fileList.get(taskIndex).patchName;
			}

			public void run(int taskIndex) throws Throwable {
				final PatchFile task = fileList.get(taskIndex);
				try {
					PatchBuilder cons = new PatchBuilder();
					PatchFile303 reader = new PatchFile303(task.fileName, cons);
					reader.readAll();
					
					final PatchUI tab = PatchUI.newInstance(cons.getPatch());
					
					/**
					 * Because Swing's not thread save.
					 */
					SwingUtilities.invokeAndWait(
						new Runnable() {
							public void run() {
								nomad.getDocumentManager().addDocument(task.patchName, tab);
							}
						}
					);
				} catch (FileNotFoundException e) {
					throw e;
				} catch (PatchConstructionException e) {
					throw e;
				} catch (Throwable t) {
					System.out.println("Error while loading patch.");
					t.printStackTrace();
				}

				if (taskIndex==getTaskCount()-1) {
					if (selection>=0 && selection<nomad.getDocumentManager().getDocumentCount())
						nomad.getDocumentManager().setSelectedDocument(selection);
					fileList.clear();
				}
			}
			
		});
		
		dlg.invoke();
		
	}
		
}

File: patchloader/src/patchloader.cc
/*
    Nord Modular patch loader
    Copyright (C) 2003 Marcus Andersson

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <stdio.h>
#include <unistd.h>

#include "nmprotocol/mididriver.h"
#include "nmprotocol/midiexception.h"
#include "nmprotocol/patchmessage.h"
#include "nmprotocol/nmprotocol.h"

#include "pdl/pdlexception.h"

#include "nmpatch/patch.h"
#include "nmpatch/patchexception.h"

void load(string, string, int, string, string, string);

extern char *optarg;
extern int optind;

int main(int argc, char** argv)
{
  printf("Patchloader version %s, Copyright (C) 2003 Marcus Andersson\n"
	 "Patchloader comes with ABSOLUTELY NO WARRANTY. This is free\n"
	 "software, and you are welcome to redistribute it under certain\n"
	 "conditions.\n", VERSION);

  try {
    MidiMessage::usePDLFile("/usr/local/lib/nmprotocol/midi.pdl", 0);
    PatchMessage::usePDLFile("/usr/local/lib/nmprotocol/patch.pdl", 0);

    string patchname, drivername, input, output;
    bool error = false;
    int slot = 0;

    int option;
    while ((option = getopt(argc, argv, "n:s:d:i:o:")) != -1) {

      switch(option) {

      case 'n':
	patchname = string(optarg);
	break;
	
      case 'd':
	drivername = string(optarg);
	break;

      case 'i':
	input = string(optarg);
	break;
	
      case 'o':
	output = string(optarg);
	break;
	
      case 's':
	sscanf(optarg, "%d", &slot);
	if (slot < 0 || slot > 3) {
	  error = true;
	}
	break;

      case '?':
	error = true;

      default:
	break;
      }
    }
    
    if (error || optind >= argc) {
      printf("usage: %s -n patchname -s {0,1,2,3} -d mididriver\n"
	     "                   -i midiinput -o midioutput patchfile\n\n",
	     argv[0]);

      printf("Available midi drivers:\n");
      MidiDriver::StringList drivers = MidiDriver::getDrivers();
      for (MidiDriver::StringList::iterator i = drivers.begin();
	   i != drivers.end(); i++) {
	printf(" %s\n", (*i).c_str());
	MidiDriver* driver = MidiDriver::createDriver(*i);
	printf("  inputs:");
	MidiDriver::StringList inputs = driver->getMidiInputPorts();
	for (MidiDriver::StringList::iterator j = inputs.begin();
	     j != inputs.end(); j++) {
	  printf(" %s", (*j).c_str());
	}
	printf("\n  outputs:");
	MidiDriver::StringList outputs = driver->getMidiOutputPorts();
	for (MidiDriver::StringList::iterator j = outputs.begin();
	     j != outputs.end(); j++) {
	  printf(" %s", (*j).c_str());
	}
	printf("\n\n");
      }
      exit(1);
    }

    printf("Loading patch...\n");
    load(argv[optind], patchname, slot, drivername, input, output);
    printf("Done.\n");

  }
  catch (MidiException& exception) {
    printf("MidiException: %s (%d)\n",
	   exception.getMessage().c_str(),
	   exception.getError());
  }
  catch (PDLException& exception) {
    printf("PDLException: %s (%d)\n",
	   exception.getMessage().c_str(),
	   exception.getError());
  }
  catch (PatchException& exception) {
    printf("PatchException: %s (%d)\n",
	   exception.getMessage().c_str(),
	   exception.getError());
  }
}

void load(string filename, string patchname, int slot,
	  string drivername, string input, string output)
{
  Patch* patch = new Patch(filename);
  patch->setName(patchname);
  PatchMessage patchMessage(patch);
  patchMessage.setSlot(slot);
  
  MidiDriver* driver = MidiDriver::createDriver(drivername);
  driver->connect(input, output);
  
  NMProtocol nmProtocol;
  nmProtocol.useMidiDriver(driver);
  
  nmProtocol.send(&patchMessage);
  
  while(!nmProtocol.sendQueueIsEmpty()) {
    nmProtocol.heartbeat();
    usleep(10000);
  }
  
  nmProtocol.useMidiDriver(0);
  driver->disconnect();
}
